File: packages/python/elizaos/features/advanced_capabilities/actions/update_contact.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_action_spec
from elizaos.prompts import UPDATE_CONTACT_TEMPLATE
from elizaos.types import (
    Action,
    ActionExample,
    ActionResult,
    Content,
    ModelType,
)
from elizaos.utils.spec_examples import convert_spec_examples
from elizaos.utils.xml import parse_key_value_xml

if TYPE_CHECKING:
    from elizaos.types import (
        HandlerCallback,
        HandlerOptions,
        IAgentRuntime,
        Memory,
        State,
    )

# Get text content from centralized specs
_spec = require_action_spec("UPDATE_CONTACT")


def _convert_spec_examples() -> list[list[ActionExample]]:
    """Convert spec examples to ActionExample format."""
    return convert_spec_examples(_spec)


@dataclass
class UpdateContactAction:
    name: str = _spec["name"]
    similes: list[str] = field(default_factory=lambda: list(_spec.get("similes", [])))
    description: str = _spec["description"]

    async def validate(
        self, runtime: IAgentRuntime, _message: Memory, _state: State | None = None
    ) -> bool:
        relationships_service = runtime.get_service("relationships")
        return relationships_service is not None

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        from elizaos.features.advanced_capabilities.services.relationships import (
            RelationshipsService,
        )

        relationships_service = runtime.get_service("relationships")
        if not relationships_service or not isinstance(relationships_service, RelationshipsService):
            return ActionResult(
                text="Relationships service not available",
                success=False,
                values={"error": True},
                data={"error": "RelationshipsService not available"},
            )

        state = await runtime.compose_state(message, ["RECENT_MESSAGES", "ENTITIES"])

        prompt = runtime.compose_prompt_from_state(
            state=state,
            template=UPDATE_CONTACT_TEMPLATE,
        )

        response = await runtime.use_model(ModelType.TEXT_SMALL, {"prompt": prompt})
        parsed = parse_key_value_xml(response)

        if not parsed or not parsed.get("contactName"):
            return ActionResult(
                text="Could not determine which contact to update",
                success=False,
                values={"error": True},
                data={"error": "No contact name provided"},
            )

        contact_name = str(parsed.get("contactName", ""))
        operation = str(parsed.get("operation", "replace"))

        contacts = await relationships_service.search_contacts(search_term=contact_name)

        if not contacts:
            return ActionResult(
                text=f"Could not find a contact named '{contact_name}'",
                success=False,
                values={"error": True},
                data={"error": "Contact not found"},
            )

        contact = contacts[0]

        categories = None
        tags = None

        if parsed.get("categories"):
            new_categories = [c.strip() for c in str(parsed["categories"]).split(",") if c.strip()]
            if operation == "add_to" and contact.categories:
                categories = list(set(contact.categories + new_categories))
            else:
                categories = new_categories

        if parsed.get("tags"):
            new_tags = [t.strip() for t in str(parsed["tags"]).split(",") if t.strip()]
            if operation == "add_to" and contact.tags:
                tags = list(set(contact.tags + new_tags))
            else:
                tags = new_tags

        updated = await relationships_service.update_contact(
            entity_id=contact.entity_id,
            categories=categories,
            tags=tags,
        )

        if updated:
            response_text = f"I've updated {contact_name}'s contact information."
            if categories:
                response_text += f" Categories: {', '.join(categories)}."
            if tags:
                response_text += f" Tags: {', '.join(tags)}."

            if callback:
                await callback(Content(text=response_text, actions=["UPDATE_CONTACT"]))

            return ActionResult(
                text=response_text,
                success=True,
                values={
                    "contactId": str(contact.entity_id),
                    "categoriesStr": ",".join(categories) if categories else "",
                    "tagsStr": ",".join(tags) if tags else "",
                },
                data={"success": True},
            )
        else:
            return ActionResult(
                text="Failed to update contact",
                success=False,
                values={"error": True},
                data={"error": "Update operation failed"},
            )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return _convert_spec_examples()


update_contact_action = Action(
    name=UpdateContactAction.name,
    similes=UpdateContactAction().similes,
    description=UpdateContactAction.description,
    validate=UpdateContactAction().validate,
    handler=UpdateContactAction().handler,
    examples=UpdateContactAction().examples,
)

File: packages/typescript/src/features/advanced-capabilities/actions/updateContact.ts
import { requireActionSpec } from "../../../generated/spec-helpers.ts";
import {
	findKeywordTermMatch,
	getValidationKeywordTerms,
} from "../../../i18n/validation-keywords.ts";
import { logger } from "../../../logger.ts";
import { updateContactTemplate } from "../../../prompts.ts";
import type {
	ContactInfo,
	RelationshipsService,
} from "../../../services/relationships.ts";
import type {
	Action,
	ActionExample,
	ActionResult,
	HandlerCallback,
	HandlerOptions,
	IAgentRuntime,
	Memory,
	State,
} from "../../../types/index.ts";
import { ModelType } from "../../../types/index.ts";
import { composePromptFromState, parseKeyValueXml } from "../../../utils.ts";

// Get text content from centralized specs
const spec = requireActionSpec("UPDATE_CONTACT");
const UPDATE_CONTACT_TERMS = getValidationKeywordTerms(
	"action.updateContact.request",
	{
		includeAllLocales: true,
	},
);

interface UpdateContactXmlResult {
	contactName?: string;
	operation?: string;
	categories?: string;
	tags?: string;
	preferences?: string;
	customFields?: string;
	notes?: string;
}

const parseKeyValueList = (value?: string): Record<string, string> => {
	if (!value) return {};
	const result: Record<string, string> = {};
	const entries = value.split(",");
	for (const entry of entries) {
		const [key, val] = entry.split(":").map((s: string) => s.trim());
		if (key && val) {
			result[key] = val;
		}
	}
	return result;
};

export const updateContactAction: Action = {
	name: spec.name,
	similes: spec.similes ? [...spec.similes] : [],
	description: spec.description,
	examples: (spec.examples ?? []) as ActionExample[][],

	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
		_state?: State,
	): Promise<boolean> => {
		const hasService = !!runtime.getService("relationships");
		const text = message.content.text;
		if (!text) return false;
		const hasIntent = findKeywordTermMatch(text, UPDATE_CONTACT_TERMS);
		return hasService && !!hasIntent;
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		_options?: HandlerOptions,
		callback?: HandlerCallback,
	): Promise<ActionResult | undefined> => {
		try {
			const relationshipsService = runtime.getService(
				"relationships",
			) as RelationshipsService;
			if (!relationshipsService) {
				throw new Error("RelationshipsService not available");
			}

			// Build state for prompt composition
			const updateState: State = {
				values: {
					...state?.values,
					message: message.content.text,
					senderName: state?.values?.senderName || "User",
					senderId: message.entityId,
					currentDateTime: new Date().toISOString(),
				},
				data: state?.data || {},
				text: state?.text || "",
			};

			const prompt = composePromptFromState({
				state: updateState,
				template: updateContactTemplate,
			});

			// Get LLM response
			const response = await runtime.useModel(ModelType.TEXT_SMALL, {
				prompt,
			});
			const parsed = parseKeyValueXml<UpdateContactXmlResult>(response);
			if (!parsed) {
				logger.warn("[UpdateContact] Failed to parse response");
				await callback?.({
					text: "I couldn't understand the update request. Please try again.",
				});
				return;
			}

			const contactName = parsed.contactName?.trim();
			if (!contactName) {
				logger.warn("[UpdateContact] No contact name provided");
				await callback?.({
					text: "I couldn't determine which contact to update. Please specify the contact name.",
				});
				return;
			}

			// Find the contact entity
			const contacts = await relationshipsService.searchContacts({
				searchTerm: contactName,
			});

			if (contacts.length === 0) {
				await callback?.({
					text: `I couldn't find a contact named "${contactName}" in the relationships.`,
				});
				return;
			}

			const contact = contacts[0];
			const operation = parsed.operation || "replace";

			// Prepare update data
			const updateData: Partial<ContactInfo> = {};

			// Handle categories
			if (parsed.categories) {
				const newCategories = parsed.categories
					.split(",")
					.map((c: string) => c.trim())
					.filter(Boolean);
				if (operation === "add_to" && contact.categories) {
					updateData.categories = [
						...new Set([...contact.categories, ...newCategories]),
					];
				} else if (operation === "remove_from" && contact.categories) {
					updateData.categories = contact.categories.filter(
						(category) => !newCategories.includes(category),
					);
				} else {
					updateData.categories = newCategories;
				}
			}

			// Handle tags
			if (parsed.tags) {
				const newTags = parsed.tags
					.split(",")
					.map((t: string) => t.trim())
					.filter(Boolean);
				if (operation === "add_to" && contact.tags) {
					updateData.tags = [...new Set([...contact.tags, ...newTags])];
				} else if (operation === "remove_from" && contact.tags) {
					updateData.tags = contact.tags.filter(
						(tag) => !newTags.includes(tag),
					);
				} else {
					updateData.tags = newTags;
				}
			}

			// Handle preferences
			if (parsed.preferences) {
				const newPrefs = parseKeyValueList(parsed.preferences);
				if (operation === "add_to" && contact.preferences) {
					updateData.preferences = { ...contact.preferences, ...newPrefs };
				} else if (operation === "remove_from" && contact.preferences) {
					const remainingPreferences = { ...contact.preferences };
					for (const key of Object.keys(newPrefs)) {
						delete remainingPreferences[key];
					}
					updateData.preferences = remainingPreferences;
				} else {
					updateData.preferences = newPrefs;
				}
			}

			// Handle custom fields
			if (parsed.customFields) {
				const newFields = parseKeyValueList(parsed.customFields);
				if (operation === "add_to" && contact.customFields) {
					updateData.customFields = { ...contact.customFields, ...newFields };
				} else if (operation === "remove_from" && contact.customFields) {
					const remainingCustomFields = { ...contact.customFields };
					for (const key of Object.keys(newFields)) {
						delete remainingCustomFields[key];
					}
					updateData.customFields = remainingCustomFields;
				} else {
					updateData.customFields = newFields;
				}
			}

			// Update the contact
			const updated = await relationshipsService.updateContact(
				contact.entityId,
				updateData,
			);

			if (updated) {
				const responseText = `I've updated ${contactName}'s contact information. ${
					updateData.categories
						? `Categories: ${updateData.categories.join(", ")}. `
						: ""
				}${updateData.tags ? `Tags: ${updateData.tags.join(", ")}. ` : ""}`;

				await callback?.({
					text: responseText,
					actions: ["UPDATE_CONTACT_INFO"],
				});

				logger.info(`[UpdateContact] Updated contact ${contact.entityId}`);

				return {
					success: true,
					values: {
						contactId: contact.entityId,
						categoriesStr: updateData.categories?.join(",") ?? "",
						tagsStr: updateData.tags?.join(",") ?? "",
					},
					data: {
						success: true,
						updatedFieldsStr: Object.keys(updateData).join(","),
					},
					text: responseText,
				};
			} else {
				throw new Error("Failed to update contact");
			}
		} catch (error) {
			logger.error(
				"[UpdateContact] Error:",
				error instanceof Error ? error.message : String(error),
			);
			await callback?.({
				text: "I encountered an error while updating the contact. Please try again.",
				error: error instanceof Error ? error.message : "Unknown error",
			});
		}
	},
};
